src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyIconSetCache.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using Microsoft.VisualStudio.Imaging.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies
+{
+    /// <summary>
+    /// Caches and reuses <see cref="DependencyIconSet"/> instances.
+    /// In practice dependencies use a relatively small number of distinct icon
+    /// sets; we can save considerable amounts of memory by ensuring that the same
+    /// logical sets are represented with the same instances.
+    /// </summary>
+    internal sealed class DependencyIconSetCache
+    {
+        private readonly ConcurrentDictionary<DependencyIconSet, DependencyIconSet> _iconSets = new ConcurrentDictionary<DependencyIconSet, DependencyIconSet>();
+
+        public DependencyIconSet GetOrAddIconSet(DependencyIconSet iconSet)
+        {
+            return _iconSets.GetOrAdd(iconSet, set => set);
+        }
+
+        public DependencyIconSet GetOrAddIconSet(ImageMoniker icon, ImageMoniker expandedIcon, ImageMoniker unresolvedIcon, ImageMoniker unresolvedExpandedIcon)
+        {
+            return GetOrAddIconSet(new DependencyIconSet(icon, expandedIcon, unresolvedIcon, unresolvedExpandedIcon));
```

Comment:
>new DependencyIconSet [](start = 35, length = 21)

You can avoid creating extra icon sets by having dictionary<hash, DependencyIconSet> ? Have static method GetHashCode(<icons> on the DependencyIconSet that would give you a hash to check for existence in the dictionary?

Reply:
Your implementation looks good and clean, i just thought if we could avoid unnecessary instantiations...

---
In reply to: [216084356](https://github.com/dotnet/project-system/pull/3959#discussion_r216084356) [](ancestors = 216084356)